### PR TITLE
[Bugfix] Solve Ulysses-SP sequence length not divisible by SP degree (using padding and attention mask)

### PR DESCRIPTION
**Pull request by wtomin**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose

As mentioned in #492, the current Ulysses-SP will raise an error if sequence length not divisible by SP degree. This PR solves this problem by:
1. check if sequence parallelism is enabled, and if sequence_length is not divisible by SP degree
2. if 1 is true, check if the current attenion backends support attention_mask, if not, will raise an error.
3. if 1 is true, check if ring_attention is enabled. If true, it will raise an error because ring_attention does not support attention_mask.
4.  then pad the input sequence to the target length. Only pad the img_hidden_states, leave encoder_hidden_states alone because it is not split across ranks, and generate an attention mask;
5. Pass attention mask to Attention layer, and handles it in `UlyssesParallelAttention.pre_attention`.

## Test Plan

1. divisble length  (no padding and attention_mask)

`python examples/offline_inference/text_to_image/text_to_image.py  --height 1024 --width 1024`

`python examples/offline_inference/text_to_image/text_to_image.py  --height 1024 --width 1024 --ulysses_degree 2`

2. non-divisble length  (padding and attention_mask)

`python examples/offline_inference/text_to_image/text_to_image.py --ulysses_degree 4 --height 1328 --width 1328
`
`python examples/offline_inference/text_to_image/text_to_image.py  --height 928 --width  1664 --ulysses_degree 4`

`
python examples/offline_inference/text_to_image/text_to_image.py --height 1024 --width 1536 --ulysses_degree 4 
`
## Test Result

1. divisble length (no padding and attention_mask)

This PR will not introduce overload because no padding or attention mask is applied.

| branch | height | width |  ulysses_degree | time (s)| output_image |
| --- | ---| ---| ---| --- | ---|
| main |  1024 | 1024 | 1 |  22.74|  qwen_image_output|
| this PR | 1024 | 1024 | 1 | 22.85 | qwen_image_output |
| main |  1024 | 1024 | 2 | 21.08 |  qwen_image_output |
| this PR | 1024 | 1024 | 2 |  21.31 |  qwen_image_output|

2.  non-divisble length (padding and attention_mask)

main branch with SP cannot deal with non-divisble length, and will lead to an error. See in #492. 

| branch | height | width | ulysses_degree | time (s)| output_image |
| --- | ---| ---| ---| --- | ---| 
| this PR | 1328 | 1328  | 4 | 31.64 | qwen_image_output|
| this PR | 928|  1664 | 4 |  25.32 | qwen_image_output | 
| this PR | 1024 |  1536  | 4 |  25.6 | qwen_image_output|

This will introduce significant overload because of attention_mask.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by hsliuustc0106** [comment]
do we need to update the docs accordingly

**Comment by david6666666** [comment]
I think we should add a test

**Comment by david6666666** [comment]
@iwzbi @ZJY0516 please review, thx

File: vllm_omni/diffusion/models/qwen_image/qwen_image_transformer.py [hunk 1]
```diff
@@ -732,14 +777,48 @@ def forward(
...
+                        f"Please switch to Ulysses SP only."
+                    )
+
+                seq_padding = sp_size - (seq_len % sp_size)
+                original_seq_len = seq_len
+
+                hidden_states_mask = torch.ones(
+                    batch_size, seq_len + seq_padding, dtype=torch.bool, device=hidden_states.device
+                )
+                hidden_states_mask[:, seq_len:] = False
```
[the remark this answers is not in the document]
Comment (wtomin) [reply vllm_omni/diffusion/models/qwen_image/qwen_image_transformer.py:810]: I don't see why the mask needs to be inverted before being sent to SDPA.

The official python implementation of [torch.nn.functional.scaled_dot_product_attention](https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html#torch.nn.functional.scaled_dot_product_attention):
```python
# Efficient implementation equivalent to the following:
def scaled_dot_product_attention(query, key, value, attn_mask=None, dropout_p=0.0,
        is_causal=False, scale=None, enable_gqa=False) -> torch.Tensor:
    ... ...
    if attn_mask is not None:
        if attn_mask.dtype == torch.bool:
            attn_bias.masked_fill_(attn_mask.logical_not(), float("-inf"))
        else:
            attn_bias = attn_mask + attn_bias
```

For the `attn_mask` passed to SDPA, only the False items (padded tokens) will be masked out with `float(-inf)`

**Comment by wtomin** [comment]
> I think we should add a test

Good idea. I think I can add an e2e test with image shape that seq_len not divisible by SP_degree.

File: vllm_omni/diffusion/models/qwen_image/qwen_image_transformer.py [hunk 2]
```diff
@@ -416,30 +425,63 @@ def forward(
...
             )
         else:
+            attn_metadata = None
+            if hidden_states_mask is not None or encoder_hidden_states_mask is not None:
+                mask_list = []
+                if encoder_hidden_states_mask is not None:
+                    mask_list.append(encoder_hidden_states_mask)
+                else:
+                    mask_list.append(
+                        torch.ones(
```
Comment (iwzbi) [inline vllm_omni/diffusion/models/qwen_image/qwen_image_transformer.py:460]: little question: why create an all True mask here

**Comment by ZJY0516** [comment]
I think we should refactor the parallel implementation later. The current approach is too invasive, requiring modifications to the model files.

**Review by ZJY0516 — APPROVED** [review]

**Comment by hsliuustc0106** [comment]
> I think we should refactor the parallel implementation later. The current approach is too invasive, requiring modifications to the model files.

agree, this is listed in Q1 roadmap, let's finish it in a follow-up PR

File: vllm_omni/diffusion/models/qwen_image/qwen_image_transformer.py (hunk 2 above)
Comment (wtomin) [reply vllm_omni/diffusion/models/qwen_image/qwen_image_transformer.py:460]: Simply a placeholder so that the two masks can be concatenated as `joint_mask`
